src/pages/FormPage/submit.ts

```diff
@@ -0,0 +1,190 @@
+import React, {SyntheticEvent} from "react";
+import * as Sentry from "@sentry/react";
+
+import RenderedQuestion from "../../components/Question";
+import {RefMapType} from "./FormPage";
+
+import ApiClient from "../../api/client";
+import {Question, QuestionType, UnittestFailure} from "../../api/question";
+import {checkScopes, OAuthScopes} from "../../api/auth";
+
+
+export enum FormState {
+    INITIAL = "initial",
+    SENDING = "sending",
+    SENT = "sent",
+    VALIDATION_ERROR = "validation_error",
+    UNKNOWN_ERROR = "error",
+}
+
+
+/**
+ * Handle validation and submission of a form.
+ *
+ * @param event The submission event.
+ * @param formID The form ID.
+ * @param questions A list of :RenderedQuestion: elements.
+ * @param refMap A map of question ID to object refs.
+ * @param setState A consumer which marks the current state of the form.
+ * @param OAuthRef A reference to the OAuth button to scroll to if the user is not logged in.
+ * @param scopes The OAuth scopes required to submit the form.
+ */
+export default async function handleSubmit(
+    event: SyntheticEvent,
+    formID: string,
+    questions: RenderedQuestion[],
+    refMap: RefMapType,
+    setState: (state: string) => void,
+    OAuthRef: React.RefObject<HTMLDivElement>,
+    scopes: OAuthScopes[]
+): Promise<void> {
+    try {
+        event.preventDefault();
+
+        if (scopes.length && !checkScopes(scopes)) {
+            // The form requires certain scopes, but the user is not logged in
+            if (!OAuthRef.current) {
+                Sentry.captureMessage("OAuthRef was not set, could not scroll to the button.");
+            } else {
+                OAuthRef.current.scrollIntoView({behavior: "smooth", block: "end"});
+            }
+
+            return;
+        }
+
+        if (!validate(questions, refMap)) {
+            return;
+        }
+
+        // FIXME: Save state while sending
+        // setState(FormState.SENDING);
+
+        await ApiClient.post(`forms/submit/${formID}`, {response: parseAnswers(questions)})
+            .then(() => setState(FormState.SENT))
+            .catch(error => {
+                if (!error.response) {
+                    throw error;
+                }
+
+                switch (error.response.status) {
+                    case 422:
+                        // TODO: Re-enable this once we have better state management
+                        // setState(FormState.VALIDATION_ERROR);
+                        showUnitTestFailures(refMap, error.response.data);
```

Comment:
Is the idea that in the future, this function should just set a state, and something consuming that state will be responsible for doing `showUnitTestFailures`?

Reply:
That's one option, but mainly this is going to be important once we're switching between the sending state, and some other state. *Some* state change is going to be necessary.

Validation error can be merged into the initial state, but I set it to its own type here in case it proves useful for something else (such as moving the validation logic, or rewriting the scroll logic).